src/keycloak/keycloak.service.ts

```diff
@@ -0,0 +1,222 @@
+import {
+  Injectable,
+  InternalServerErrorException,
+  Logger,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { ok } from 'assert';
+import { HttpClientService } from '../httpclient/httpclient.service';
+import { KeyCloakConfigProperties } from './keycloak.config.properties';
+import jwkToPem, { JWK } from 'jwk-to-pem';
+import { stringify } from 'qs';
+import { verify } from 'jsonwebtoken';
+
+export interface OIDCIdentityConfig {
+  issuer?: string;
+  introspection_endpoint?: string;
+  jwks_uri?: string;
+  token_endpoint?: string;
+}
+
+export interface RegisterUserData {
+  email: string;
+  firstName: string;
+  lastName: string;
+  password: string;
+}
+
+export interface GenerateTokenData {
+  username: string;
+  password: string;
+}
+
+export interface Token {
+  readonly token_type: string;
+  readonly access_token: string;
+  readonly refresh_token?: string;
+  readonly refresh_expires_in: number;
+  readonly expires_in?: number;
+}
+
+interface ClientCredentials {
+  client_id: string;
+  client_secret?: string;
+}
+
+@Injectable()
+export class KeyCloakService {
+  private log: Logger = new Logger(KeyCloakService.name);
+  private readonly server_uri: string;
+  private readonly realm: string;
+  private config: OIDCIdentityConfig;
+  private readonly clientAdmin: ClientCredentials;
+  private readonly clientUser: ClientCredentials;
+
+  constructor(
+    private readonly configService: ConfigService,
+    private readonly httpClient: HttpClientService,
+  ) {
+    this.server_uri = this.configService.get(
+      KeyCloakConfigProperties.ENV_KEYCLOAK_SERVER_URI,
+    );
+    this.realm = this.configService.get(
+      KeyCloakConfigProperties.ENV_KEYCLOAK_REALM,
+    );
+
+    this.clientUser = {
+      client_id: this.configService.get(
+        KeyCloakConfigProperties.ENV_KEYCLOAK_USER_CLIENT_ID,
+      ),
+      client_secret: this.configService.get(
+        KeyCloakConfigProperties.ENV_KEYCLOAK_USER_CLIENT_SECRET,
+      ),
+    };
+
+    this.clientAdmin = {
+      client_id: this.configService.get(
+        KeyCloakConfigProperties.ENV_KEYCLOAK_ADMIN_CLIENT_ID,
+      ),
+      client_secret: this.configService.get(
+        KeyCloakConfigProperties.ENV_KEYCLOAK_ADMIN_CLIENT_SECRET,
+      ),
+    };
+
+    ok(this.realm, '"realm" is not defined.');
+    ok(this.server_uri, '"metadata_uri" is not defined.');
+    ok(this.clientUser.client_id, '"client_id" is not defined.');
+    ok(this.clientAdmin.client_id, 'Admin "client_id" is not defined.');
+    ok(this.clientAdmin.client_secret, 'Admin "client_secret" is not defined.');
+
+    this.discovery().catch((err: any) => console.error(err));
```

Comment:
Why don't use `OnModuleInit` hook or async provider? You should avoid calling long operations from the constructor (async or sync)